plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,524 @@
+	for _, entity := range clusterCmmds.Res {
+		cmmdsMap[entity.UUID] = entity
+	}
+	return cmmdsMap, nil
+}
+
+// queryPerformance adds performance metrics to telegraf accumulator
+func (e *Endpoint) queryPerformance(ctx context.Context, vsanClient *soap.Client, clusterRef *objectRef, metrics map[string]string,
+	cmmds map[string]CmmdsEntity, acc telegraf.Accumulator) error {
+	end := time.Now().UTC()
+
+	// We're using a fake metric key, since we only store one highwater mark per resource
+	start, ok := e.hwMarks.Get(hwMarksKeyPrefix+clusterRef.ref.Value, "generic")
+	if !ok {
+		// Look back 3 sampling periods by default
+		start = end.Add(time.Duration(e.Parent.MetricLookback) * time.Duration(-e.resourceKinds["vsan"].sampling) * time.Second)
+	}
+	e.Parent.Log.Debugf("[vSAN] Query vsan performance for time interval: %s ~ %s", start, end)
+	latest := start
+
+	var commonError error
+	for entityRefID := range metrics {
+		if !strings.HasPrefix(entityRefID, perfPrefix) {
+			continue
+		}
+		entityRefID = strings.TrimPrefix(entityRefID, perfPrefix)
+		var perfSpecs []vsantypes.VsanPerfQuerySpec
+
+		perfSpec := vsantypes.VsanPerfQuerySpec{
+			EntityRefId: entityRefID + ":*",
+			StartTime:   &start,
+			EndTime:     &end,
+		}
+		perfSpecs = append(perfSpecs, perfSpec)
+
+		perfRequest := vsantypes.VsanPerfQueryPerf{
+			This:       perfManagerRef,
+			QuerySpecs: perfSpecs,
+			Cluster:    &clusterRef.ref,
+		}
+		resp, err := vsanmethods.VsanPerfQueryPerf(ctx, vsanClient, &perfRequest)
+		if err != nil {
+			if err.Error() == "ServerFaultCode: NotFound" {
+				e.Parent.Log.Errorf("[vSAN] Is vSAN performance service enabled for %s? Skipping ...", clusterRef.name)
+				commonError = err
+				break
+			}
+			e.Parent.Log.Errorf("[vSAN] Error querying performance data for %s: %s: %s.", clusterRef.name, entityRefID, err)
+			continue
+		}
+		tags := populateClusterTags(make(map[string]string), clusterRef, e.URL.Host)
+
+		count := 0
+		for _, em := range resp.Returnval {
+			vals := strings.Split(em.EntityRefId, ":")
+			var entityName, uuid string
+			if len(vals) == 1 {
+				entityName, uuid = vals[0], ""
+			} else {
+				entityName, uuid = vals[0], vals[1]
+			}
+
+			buckets := make(map[string]metricEntry)
+			tags := populateCMMDSTags(tags, entityName, uuid, cmmds)
+			var timeStamps []time.Time
+			// 1. Construct a timestamp list from sample info
+			formattedEntityName := hyphenReplacer.Replace(entityName)
+			for _, t := range strings.Split(em.SampleInfo, ",") {
+				tsParts := strings.Split(t, " ")
+				if len(tsParts) >= 2 {
+					// The return time string is in UTC time
+					utcTimeStamp := fmt.Sprintf("%sT%sZ", tsParts[0], tsParts[1])
+					ts, ok := time.Parse(time.RFC3339, utcTimeStamp)
+					if ok != nil {
+						e.Parent.Log.Errorf("[vSAN] Failed to parse a timestamp: %s. Skipping", utcTimeStamp)
+						timeStamps = append(timeStamps, time.Time{})
+						continue
+					}
+					timeStamps = append(timeStamps, ts)
+				}
```

Comment:
Why don't you specify your custom time format directly instead of splitting and combining stuff?